### PR TITLE
Changes ids_get() Default Parameters for counterparts and start_date to Reduce Data Volume and Align with User Needs #50

### PR DESCRIPTION
**Pull request by t-emery**
Changed default parameters for `start_date` and `counterparts` for `ids_get()`. 
- Improved function documentation & examples, partially addressing #47.
- changed date behavior.  Initially despite changing `start_date` to 2000, it would still provide data from 1970 if end_date was empty (it worked as expected when end_date was specified).  I fixed this, and created filtering logic so that debt service data with projections through 2031 would show up, but for data without projections it filtered out NA values after 2023 (last observed data)
- Fixing this required using hard-coded dates.  These can be updated every year (Perhaps we can document the "yearly data update checklist in the wiki?   Another option would be to add a `ids_list_years()` and create some logic that pulled the years from there automatically.  I figured we have more pressing issues, but might want to address that later. 

This is my first pull request!  Any feedback welcome.   
-

### CODE REVIEW

File: R/ids_get.R
Comment (chriscarrollsmith) [inline R/ids_get.R:30]: Important to be careful about the terms "numeric" and "text" since these have a special meaning in R. (I.e., are we supposed to provide 730 as a numeric variable but "907" as a string?)

File: R/ids_get.R
Comment (chriscarrollsmith) [inline R/ids_get.R:40]: We actually could remove this constraint, replace any sub-1970 year with 1970, and just raise a warning flag to let the user know we did so.

File: R/ids_get.R
Comment (chriscarrollsmith) [inline R/ids_get.R:35]: I do think that if you set the default start date to 2000, you risk misleading casual users into thinking no earlier data is available.

File: R/ids_get.R
Comment (chriscarrollsmith) [inline R/ids_get.R:269]: I take it you're trying to drop NAs outside the coverage period, but preserve NAs within the coverage period? To avoid hardcoding and continually updating the start and end years, I would suggest:

1) Get the first and last years in the retrieved dataset with a non-NA value, and treat these as the boundaries of the data, then
2) Drop all NAs outside those boundaries.

**Review by chriscarrollsmith — APPROVED** [review]

**Comment by chriscarrollsmith** [comment]
I pushed a fix for a typo and a missing mock that were breaking some of the tests. Approved the PR but also added some comments above on the `ids_get` documentation that you should take a look at before merging. If you'd like me to tackle any additional adjustments, let me know.

**Comment by chriscarrollsmith** [comment]
Closes #50

File: R/ids_get.R
Comment (christophscheuch) [inline R/ids_get.R:11]: If you use \link{ids_list_geographies} instead of `ids_list_geographies()`, then you get a hyperlink to the function in the docs and I believe this is desired :)

File: R/ids_get.R
Comment (christophscheuch) [reply R/ids_get.R:30]: Yes, we have to provide identifiers as string (I guess because there can also be leading 0s?). I also believe that we should just write about text codes even if they are numbers.

**Review by christophscheuch — CHANGES_REQUESTED** [review]
Looks good mostly! There is a test coverage gap now according to the codecov report: the multi-page request is not fired during testing because we don't pull enough data through ids_get() - need to include at least one test / example where we download a lot of data from the API.

**Comment by chriscarrollsmith** [comment]
Only remaining outstanding issue here is those two hard-coded year variables, which I'd like to either:

1) get programmatically so we don't forget to update them manually, or
2) avoid altogether by handling NAs differently.

Will have a chat with Teal about that to make a design decision and then hopefully get this merged shortly thereafter.

**Comment by t-emery** [comment]
We can get the end of the debt service projection years (2031 currently, I believe) from the Time API endpoint (not sure if that is the correct terminology). Could implement a ids_list_years() to do that. And then could simplify the logic by assuming 8 years of projections (currently observed data ends in 2023).  That is one way to do this. But maybe there is an even easier way.Sent from my iPhoneOn Dec 12, 2024, at 11:37 PM, Christopher Carroll Smith ***@***.***> wrote:﻿
Only remaining outstanding issue here is those two hard-coded year variables, which I'd like to either:

get programmatically so we don't forget to update them manually, or
avoid altogether by handling NAs differently.

Will have a chat with Teal about that to make a design decision and then hopefully get this merged shortly thereafter.

**Comment by christophscheuch** [comment]
> We can get the end of the debt service projection years (2031 currently, I believe) from the Time API endpoint (not sure if that is the correct terminology). Could implement a ids_list_years() to do that. And then could simplify the logic by assuming 8 years of projections (currently observed data ends in 2023).  That is one way to do this. But maybe there is an even easier way.

Indeed, the time points are here: https://api.worldbank.org/v2/sources/6/time?per_page=1000&format=JSON. If you build a new function that fetches this, we should use `ids_get_time()` (or _dates, or _years?) because `list` refers to stuff that is fixed and shipped with the package.

**Comment by chriscarrollsmith** [comment]
Just chatted with Teal. I will build a list function as a raw data script,
and we will ship the years list with the package. (Will probably also need
to build in an automation to make sure this runs when we build the package,
so it's regularly updated.)

**Comment by chriscarrollsmith** [comment]
I think I've got this done. @christophscheuch, you want to review and pull trigger?

**Review by christophscheuch — APPROVED** [review]

**Pull request merged**
